File: frame.py
"""
HAVEN-FSK Frame Builder
=========================
Assembles and disassembles complete over-the-air frames.

Frame structure (transmitted order):
    PREAMBLE   16 MFSK symbols — timing recovery and signal ID
               Fixed alternating pattern: tones [0,15,0,15,7,8,7,8,...]
               Receiver correlates against this to find frame start
               and recover exact symbol timing.

    HEADER     2 bytes (16 bits) — transmitted before FEC payload
        Byte 0: VERSION (4 bits) | FLAGS (4 bits)
                VERSION = 0001 for this release
                FLAGS bit 0: 1 = FEC enabled, 0 = FEC disabled
                FLAGS bit 1: 1 = more frames follow (fragmented)
                FLAGS bits 2-3: reserved
        Byte 1: NBLOCKS (8 bits) — number of FEC blocks in payload
                Limits messages to 255 * 12 = 3060 bytes max
                Well beyond any practical message length

    CRC-16     2 bytes — CRC of (HEADER + PAYLOAD)
               Polynomial: 0x1021 (CRC-CCITT)
               Receiver verifies after FEC decode
               Failed CRC triggers ⚠ CRC Error display

    PAYLOAD    NBLOCKS * 192 bits of FEC-encoded data
               Each 192-bit block carries 96 bits of message
               See fec.py for LDPC(192,96) specification

    [carrier drop = end of frame — no end marker transmitted]
    [trailing space in payload ensures last symbol fully received]

The header and CRC are transmitted WITHOUT FEC protection because:
  1. They are short (4 bytes total)
  2. The preamble correlation already provides high confidence
  3. FEC is applied to the payload where it matters most
  4. Keeping header unprotected reduces complexity

If header is corrupted the frame fails silently (DCD holdoff
prevents most header corruption in practice).
"""

import numpy as np
import struct

# ── CRC-16 CCITT ─────────────────────────────────────────────────────────────

CRC16_POLY = 0x1021

def crc16(data: bytes) -> int:
    """
    CRC-16/CCITT-FALSE
    Initial value: 0xFFFF
    Polynomial:    0x1021
    Input/output reflect: False
    """
    crc = 0xFFFF
    for byte in data:
        crc ^= byte << 8
        for _ in range(8):
            if crc & 0x8000:
                crc = (crc << 1) ^ CRC16_POLY
            else:
                crc <<= 1
            crc &= 0xFFFF
    return crc


def crc16_bytes(data: bytes) -> bytes:
    """Return CRC-16 as 2 big-endian bytes."""
    return struct.pack('>H', crc16(data))


def crc16_check(data: bytes, crc_bytes: bytes) -> bool:
    """Verify CRC. Returns True if valid."""
    try:
        expected = struct.unpack('>H', crc_bytes)[0]
        return crc16(data) == expected
    except Exception:
        return False


# ── Frame constants ───────────────────────────────────────────────────────────

VERSION      = 0b0001       # protocol version 1
FLAG_FEC     = 0b00000001   # FEC enabled
FLAG_MORE    = 0b00000010   # more frames follow (future use)

HEADER_LEN   = 2            # bytes
CRC_LEN      = 2            # bytes
MAX_BLOCKS   = 255          # max FEC blocks per frame
MAX_MSG_BYTES = MAX_BLOCKS * 12  # 3060 bytes


# ── Frame builder ─────────────────────────────────────────────────────────────

def build_frame(text: str, use_fec: bool = True) -> tuple:
    """
    Build a complete transmit frame from a text message.

    Returns:
        header:   2 bytes
        crc:      2 bytes
        payload:  encoded bytes (FEC or raw)
        n_blocks: number of FEC blocks (0 if no FEC)
        orig_len: original message byte length
    """
    raw = text.encode('utf-8')

    # Limit message length
    if len(raw) > MAX_MSG_BYTES:
        raw = raw[:MAX_MSG_BYTES]

    if use_fec:
        from fec import encode_message, BYTES_PER_BLOCK, N as FEC_N
        # Encode through FEC — returns BPSK float array
        coded_floats, n_blocks, orig_len = encode_message(text)
        # Convert back to bytes for CRC calculation
        # We CRC the original payload not the FEC output
        flags   = FLAG_FEC
        payload = raw   # CRC covers original message
    else:
        n_blocks = 0
        orig_len = len(raw)
        flags    = 0
        payload  = raw

    # Build header
    header = bytes([
        ((VERSION & 0x0F) << 4) | (flags & 0x0F),
        n_blocks & 0xFF,
    ])

    # CRC covers header + original payload
    crc = crc16_bytes(header + payload)

    if use_fec:
        return header, crc, coded_floats, n_blocks, orig_len
    else:
        return header, crc, payload, n_blocks, orig_len


def parse_header(header_bytes: bytes) -> tuple:
    """
    Parse 2-byte header.
    Returns (version, flags, n_blocks) or None on error.
    """
    if len(header_bytes) < HEADER_LEN:
        return None
    try:
        byte0, byte1 = header_bytes[0], header_bytes[1]
        version  = (byte0 >> 4) & 0x0F
        flags    = byte0 & 0x0F
        n_blocks = byte1
        return version, flags, n_blocks
    except Exception:
        return None


def verify_frame(header: bytes,
                 crc: bytes,
                 decoded_text: str) -> bool:
    """
    Verify CRC after decode.
    Returns True if frame integrity check passes.
    """
    payload = decoded_text.encode('utf-8')
    return crc16_check(header + payload, crc)


# ── Frame assembler for transmission ─────────────────────────────────────────

class FrameAssembler:
    """
    High-level frame assembly for the transmit path.
    Takes a text message, returns audio samples ready to transmit.
    The modulator handles preamble and symbol generation.
    """

    def __init__(self, use_fec: bool = True):
        self.use_fec = use_fec

    def assemble(self, text: str) -> dict:
        """
        Assemble a complete frame.

        Returns dict with:
            header:       bytes
            crc:          bytes
            payload:      bytes or float32 array
            n_blocks:     int
            orig_len:     int
            use_fec:      bool
            text:         original text
            char_count:   character count
        """
        # Add trailing space to ensure last symbol fully transmits
        # The space is stripped on receive by .strip()
        padded_text = text + ' '

        header, crc, payload, n_blocks, orig_len = build_frame(
            padded_text, use_fec=self.use_fec)

        return {
            'header':     header,
            'crc':        crc,
            'payload':    payload,
            'n_blocks':   n_blocks,
            'orig_len':   orig_len,
            'use_fec':    self.use_fec,
            'text':       text,
            'char_count': len(text),
        }


class FrameParser:
    """
    High-level frame parsing for the receive path.
    Takes decoded symbols/audio and returns verified text.
    """

    def __init__(self):
        self._pending_header = None
        self._pending_crc    = None

    def parse(self,
              audio: np.ndarray,
              snr_estimate: float = 5.0) -> dict:
        """
        Parse a received frame from accumulated audio.

        Returns dict with:
            text:        decoded text (may contain U+FFFD for corrupt chars)
            crc_ok:      True if CRC passed
            n_blocks:    number of FEC blocks decoded
            version:     protocol version
            use_fec:     whether FEC was used
            error:       error message if parsing failed, else None
        """
        from modem import (Demodulator, SAMPLES_PER_SYMBOL,
                           BASE_FREQ, SYMBOL_RATE, NUM_TONES)
        from fec import decode_message, N as FEC_N, BYTES_PER_BLOCK

        # The audio starts at the beginning of the header
        # (preamble has already been stripped by the sync layer)

        demod = Demodulator()
        sps   = SAMPLES_PER_SYMBOL

        # Extract symbols
        n_syms = len(audio) // sps
        syms   = []
        soft   = []

        for i in range(n_syms):
            block    = audio[i*sps:(i+1)*sps]
            sym, eng = demod._detect_symbol(block)
            syms.append(sym)
            soft.append(eng)

        if len(syms) < 4:
            return self._error('Frame too short')

        # Decode header (first 4 symbols = 2 bytes, unprotected)
        # Each byte = 2 symbols at 4 bits per symbol
        header_syms = syms[:4]
        header_bytes = demod.symbols_to_bytes(header_syms)
        if len(header_bytes) < HEADER_LEN:
            return self._error('Header too short')

        header = header_bytes[:HEADER_LEN]
        parsed = parse_header(header)
        if parsed is None:
            return self._error('Invalid header')

        version, flags, n_blocks = parsed

        if version != VERSION:
            return self._error(f'Unknown version {version}')

        use_fec = bool(flags & FLAG_FEC)

        # Decode CRC (next 4 symbols = 2 bytes)
        crc_syms  = syms[4:8]
        crc_bytes = demod.symbols_to_bytes(crc_syms)
        if len(crc_bytes) < CRC_LEN:
            return self._error('CRC too short')
        crc = crc_bytes[:CRC_LEN]

        # Decode payload
        payload_audio = audio[8 * sps:]

        if use_fec and n_blocks > 0:
            # FEC decode
            # Convert audio to float and pass to FEC decoder
            if len(payload_audio) < n_blocks * FEC_N * sps // 4:
                return self._error('Payload truncated')

            # Build received signal from soft symbol energies
            payload_syms = syms[8:]
            if len(payload_syms) < n_blocks * FEC_N // 4:
                return self._error('Insufficient payload symbols')

            # Reconstruct soft LLR values from symbol energies
            # Each FEC bit maps to a BPSK symbol via the tone selection
            # Use the raw audio for LLR computation
            received  = np.array(payload_audio, dtype=np.float32)
            # Normalize
            peak = np.max(np.abs(received))
            if peak > 0:
                received = received / peak

            text = decode_message(received, n_blocks,
                                   n_blocks * BYTES_PER_BLOCK,
                                   snr_estimate)
            text = text.rstrip('\x00\n\r\t').strip()

        else:
            # No FEC — raw decode
            payload_syms  = syms[8:]
            payload_bytes = demod.symbols_to_bytes(payload_syms)
            try:
                text = payload_bytes.decode('utf-8', errors='replace')
            except Exception:
                text = payload_bytes.decode('latin-1', errors='replace')
            text = text.rstrip('\x00\n\r\t').strip()

        # Verify CRC
        crc_ok = verify_frame(header, crc, text)

        return {
            'text':     text,
            'crc_ok':   crc_ok,
            'n_blocks': n_blocks,
            'version':  version,
            'use_fec':  use_fec,
            'error':    None,
        }

    def _error(self, msg: str) -> dict:
        return {
            'text':     None,
            'crc_ok':   False,
            'n_blocks': 0,
            'version':  None,
            'use_fec':  False,
            'error':    msg,
        }


# ── Self-test ─────────────────────────────────────────────────────────────────

if __name__ == '__main__':
    print("Testing CRC-16...")
    # Known CRC test vector
    test_data = b'123456789'
    crc       = crc16(test_data)
    print(f"  CRC of {test_data}: 0x{crc:04X} (expect 0x29B1)")
    assert crc == 0x29B1, f"CRC wrong: 0x{crc:04X}"
    print("  CRC test PASSED")
    print()

    print("Testing frame assembly...")
    asm = FrameAssembler(use_fec=False)

    for msg in ['CQ POTA DE WD9N K-1234 K',
                'KC8TYK DE WD9N UR 599 K',
                'Hi Tim, great signal today 73']:
        frame = asm.assemble(msg)
        print(f"  '{msg}'")
        print(f"    Header:  {frame['header'].hex()}")
        print(f"    CRC:     {frame['crc'].hex()}")
        print(f"    Payload: {len(frame['payload'])} bytes")
        print(f"    FEC:     {frame['use_fec']}")

        # Verify CRC
        ok = verify_frame(frame['header'], frame['crc'],
                          (msg + ' '))   # +space from padded_text
        print(f"    CRC OK:  {ok}")
        print()

    print("All frame tests PASSED")

File: cpp/src/dsp/Frame.cpp
#include "Frame.h"
#include "FEC.h"
#include "Modulator.h"
#include "Demodulator.h"
#include "Preamble.h"
#include "Interleaver.h"
#include <algorithm>
#include <cstring>
#include <cassert>
#include <QDebug>
#include <QString>

namespace HavenFSK {

Frame::Frame() {}

// ── CRC-16/CCITT-FALSE ────────────────────────────────────────────────────
// Polynomial: 0x1021, Initial: 0xFFFF, No reflection
// Test vector: crc16("123456789") == 0x29B1

uint16_t Frame::crc16(const uint8_t* data, size_t len) {
    uint16_t crc = CRC16_INIT;  // 0xFFFF from Constants.h
    for (size_t i = 0; i < len; i++) {
        crc ^= static_cast<uint16_t>(data[i]) << 8;
        for (int b = 0; b < 8; b++) {
            if (crc & 0x8000)
                crc = (crc << 1) ^ CRC16_POLY;  // 0x1021 from Constants.h
            else
                crc <<= 1;
        }
    }
    return crc;
}

uint16_t Frame::crc16(const std::vector<uint8_t>& data) {
    return crc16(data.data(), data.size());
}

// ── Private helpers ───────────────────────────────────────────────────────

std::array<uint8_t, 2> Frame::buildHeader(uint8_t nBlocks) const {
    uint8_t byte0 = static_cast<uint8_t>(
        ((PROTOCOL_VERSION & 0x0F) << 4) | (FLAG_FEC_ENABLED & 0x0F));
    return {byte0, nBlocks};
}

bool Frame::parseHeader(const std::array<uint8_t, 2>& hdr,
                        uint8_t& version,
                        uint8_t& flags,
                        uint8_t& nBlocks) const {
    version = (hdr[0] >> 4) & 0x0F;
    flags   = hdr[0] & 0x0F;
    nBlocks = hdr[1];
    return (version == PROTOCOL_VERSION);
}

std::array<uint8_t, 2> Frame::majorityVoteHeader(
    const std::vector<std::array<uint8_t, 2>>& copies)
{
    std::array<uint8_t, 2> result{0, 0};
    int n = static_cast<int>(copies.size());
    for (int byteIdx = 0; byteIdx < 2; byteIdx++) {
        uint8_t out = 0;
        for (int bit = 7; bit >= 0; bit--) {
            int ones = 0;
            for (const auto& copy : copies)
                if ((copy[byteIdx] >> bit) & 1) ones++;
            int bitVal = (ones * 2 >= n) ? 1 : 0;
            out = static_cast<uint8_t>((out << 1) | bitVal);
        }
        result[byteIdx] = out;
    }
    return result;
}

int Frame::argmax(const std::vector<float>& energies) {
    return static_cast<int>(
        std::distance(energies.begin(),
                      std::max_element(energies.begin(), energies.end())));
}

std::vector<uint8_t> Frame::hardDecodeSymbols(
    const std::vector<std::vector<float>>& syms,
    int offset,
    int count)
{
    // count = number of symbols to process (must be even)
    // Returns count/2 bytes
    std::vector<uint8_t> bytes;
    bytes.reserve(count / 2);
    for (int i = 0; i + 1 < count; i += 2) {
        // Gray-decode tone index back to nibble value — mirrors grayEncode() in Modulator.
        int s0 = grayDecode(static_cast<uint8_t>(argmax(syms[offset + i])));
        int s1 = grayDecode(static_cast<uint8_t>(argmax(syms[offset + i + 1])));
        bytes.push_back(static_cast<uint8_t>((s0 << 4) | (s1 & 0x0F)));
    }
    return bytes;
}

std::vector<uint8_t> Frame::bpskToBytes(const std::vector<float>& bpsk) {
    // +1.0f -> 0, -1.0f -> 1
    int nBits = static_cast<int>(bpsk.size());
    std::vector<uint8_t> bits(nBits);
    for (int i = 0; i < nBits; i++)
        bits[i] = (bpsk[i] < 0.0f) ? 1 : 0;
    return FEC::packBits(bits);
}

std::string Frame::stripPadding(const std::string& s) {
    // Strip trailing null bytes
    size_t end = s.size();
    while (end > 0 && s[end - 1] == '\0') end--;
    std::string trimmed = s.substr(0, end);
    // Strip exactly one trailing space — the padding space added by assemble().
    // Removing all whitespace would corrupt messages with intentional trailing
    // spaces and break CRC verification on the RX side.
    if (!trimmed.empty() && trimmed.back() == ' ')
        trimmed.pop_back();
    return trimmed;
}

// ── Transmit path ─────────────────────────────────────────────────────────

std::vector<float> Frame::assemble(const std::string& text) const {
    // 1. Append trailing space — ensures last symbol fully transmitted
    std::string padded = text + " ";
    std::vector<uint8_t> payload(padded.begin(), padded.end());

    // 2. FEC encode
    FEC fec;
    auto enc = fec.encodeMessage(payload);

    // 3. Build header
    auto hdr = buildHeader(static_cast<uint8_t>(enc.nBlocks));

    // 4. CRC covers header + ORIGINAL unencoded payload bytes
    std::vector<uint8_t> crcInput;
    crcInput.reserve(2 + payload.size());
    crcInput.push_back(hdr[0]);
    crcInput.push_back(hdr[1]);
    crcInput.insert(crcInput.end(), payload.begin(), payload.end());
    uint16_t crcVal = crc16(crcInput);

    // 5. CRC as 2 big-endian bytes
    std::array<uint8_t, 2> crcBytes = {
        static_cast<uint8_t>((crcVal >> 8) & 0xFF),
        static_cast<uint8_t>(crcVal & 0xFF)
    };

    // 6. Modulate each section
    Preamble preamble;
    Modulator mod;

    // Generate preamble with continuous phase accumulator
    std::vector<float> preambleAudio = preamble.generate();

    // Seed Modulator with preamble's final phase — header starts exactly
    // where preamble left off. Zero discontinuity at preamble→header boundary.
    mod.setPhase(preamble.finalPhase());

    // Header sent 3x — each copy is continuous-phase from the previous.
    // RX majority-votes the 3 copies bit-by-bit (see Frame::parse()),
    // tolerating a fade hitting up to one full copy without corrupting
    // the recovered header.
    std::vector<uint8_t> hdrVec(hdr.begin(), hdr.end());
    std::vector<float> headerAudio = mod.modulate(hdrVec);
    for (int copy = 1; copy < 3; ++copy) {
        std::vector<float> headerAudioN = mod.modulate(hdrVec);
        headerAudio.insert(headerAudio.end(),
                           headerAudioN.begin(), headerAudioN.end());
    }

    std::vector<uint8_t> crcVec(crcBytes.begin(), crcBytes.end());
    std::vector<float> crcAudio = mod.modulate(crcVec);

    // Interleave payload bits across all nBlocks LDPC blocks before
    // packing to bytes/symbols — spreads HF fade bursts so LDPC sees
    // scattered errors rather than a concentrated run (see Interleaver.h).
    std::vector<float> interleavedBpsk = Interleaver::interleave(enc.bpsk, enc.nBlocks);
    std::vector<uint8_t> payloadBytes = bpskToBytes(interleavedBpsk);
    std::vector<float> payloadAudio = mod.modulate(payloadBytes);

    // 7. Concatenate
    std::vector<float> frame;
    frame.reserve(preambleAudio.size() + headerAudio.size() +
                  crcAudio.size()     + payloadAudio.size());
    frame.insert(frame.end(), preambleAudio.begin(), preambleAudio.end());
    frame.insert(frame.end(), headerAudio.begin(),   headerAudio.end());
    frame.insert(frame.end(), crcAudio.begin(),      crcAudio.end());
    frame.insert(frame.end(), payloadAudio.begin(),  payloadAudio.end());

    return frame;
}

// ── Receive path ──────────────────────────────────────────────────────────

ParseResult Frame::parse(
    const std::vector<std::vector<float>>& softSymbols) const
{
    ParseResult result{};
    result.crcOk          = false;
    result.converged      = false;
    result.nBlocks        = 0;
    result.fecIterations  = 0;
    result.version        = 0;
    result.useFec         = false;

    // Minimum: 4 header symbols + 4 CRC symbols
    if ((int)softSymbols.size() < PAYLOAD_START) {
        result.error = "Frame too short";
        return result;
    }

    // ── Decode header — 3 copies, bit-level majority vote ──────────────────
    std::vector<std::array<uint8_t, 2>> hdrCopies;
    for (int c = 0; c < HEADER_COPIES; c++) {
        auto bytes = hardDecodeSymbols(softSymbols, c * HEADER_SYMS, HEADER_SYMS);
        if ((int)bytes.size() < HEADER_BYTES) {
            result.error = "Header decode failed";
            return result;
        }
        hdrCopies.push_back({bytes[0], bytes[1]});
    }

    if (!(hdrCopies[0] == hdrCopies[1] && hdrCopies[1] == hdrCopies[2])) {
        qDebug() << "Frame::parse: header copies differ — nBlocks candidates:"
                 << hdrCopies[0][1] << hdrCopies[1][1] << hdrCopies[2][1]
                 << "— using bit-level majority vote";
    }
    std::array<uint8_t, 2> hdr = majorityVoteHeader(hdrCopies);

    qDebug() << "Frame::parse hdr[0]=0x"
                + QString::number(hdr[0], 16).rightJustified(2, '0')
                + " hdr[1]=0x"
                + QString::number(hdr[1], 16).rightJustified(2, '0')
                + " expect hdr[0]=0x21 (3x-header majority vote, PAYLOAD_START=16)";

    uint8_t version, flags, nBlocks;
    if (!parseHeader(hdr, version, flags, nBlocks)) {
        result.error = "Unknown protocol version";
        return result;
    }
    result.version = version;
    result.useFec  = (flags & FLAG_FEC_ENABLED) != 0;
    result.nBlocks = nBlocks;

    // Sanity-check nBlocks before trusting it for buffer sizing.
    // A corrupted header byte can produce nBlocks=200+, which would demand
    // thousands of seconds of audio and cause a long collection timeout.
    constexpr int MAX_NBLOCKS = 125;
    if (nBlocks > MAX_NBLOCKS) {
        result.error = "nBlocks out of range";
        return result;
    }

    // ── Decode CRC (symbols 8-11, after double header) ────────────────────
    auto crcBytes = hardDecodeSymbols(softSymbols, HEADER_TOTAL_SYMS, CRC_SYMS);
    if ((int)crcBytes.size() < CRC_BYTES) {
        result.error = "CRC decode failed";
        return result;
    }
    uint16_t rxCrc = (static_cast<uint16_t>(crcBytes[0]) << 8) | crcBytes[1];

    // ── Decode payload (symbols 8+) ───────────────────────────────────────
    if (result.useFec && nBlocks > 0) {
        int symsNeeded = nBlocks * (LDPC_N / BITS_PER_SYMBOL);  // nBlocks * 48
        if ((int)softSymbols.size() < PAYLOAD_START + symsNeeded) {
            result.error = "Payload truncated";
            return result;
        }

        // Extract payload soft symbols
        std::vector<std::vector<float>> payloadSoft(
            softSymbols.begin() + PAYLOAD_START,
            softSymbols.begin() + PAYLOAD_START + symsNeeded);

        // Convert soft symbol energies to per-bit LLR values, then undo
        // the TX-side interleave (see Frame::assemble()) before FEC
        // decode — LLRs are still in the interleaved (transmission)
        // order at this point, matching Interleaver's bit-level domain.
        FEC fec;
        auto llr = fec.softToLLR(payloadSoft);
        auto deinterleavedLlr = Interleaver::deinterleave(llr, nBlocks);

        int origLen = nBlocks * LDPC_BYTES_PER_BLOCK;
        auto decRes = fec.decodeMessage(deinterleavedLlr, nBlocks, origLen);

        result.converged     = decRes.allConverged;
        result.fecIterations = decRes.totalIterations;

        std::string text(decRes.bytes.begin(), decRes.bytes.end());
        result.text = stripPadding(text);

    } else {
        // No FEC — hard decode payload symbols directly
        int payloadSyms = static_cast<int>(softSymbols.size()) - PAYLOAD_START;
        if (payloadSyms > 0) {
            auto rawBytes = hardDecodeSymbols(softSymbols,
                                              PAYLOAD_START,
                                              payloadSyms & ~1);
            std::string text(rawBytes.begin(), rawBytes.end());
            result.text = stripPadding(text);
        }
        result.converged = true;  // No FEC = always "converged"
    }

    // ── Verify CRC ────────────────────────────────────────────────────────
    // CRC was computed over: header bytes + original padded payload bytes
    // We must recompute over header + decoded text + " " (the trailing space)
    std::string paddedText = result.text + " ";
    std::vector<uint8_t> crcInput;
    crcInput.reserve(2 + paddedText.size());
    crcInput.push_back(hdr[0]);
    crcInput.push_back(hdr[1]);
    crcInput.insert(crcInput.end(), paddedText.begin(), paddedText.end());
    uint16_t calcCrc = crc16(crcInput);

    result.crcOk  = (calcCrc == rxCrc);
    result.error  = "";

    return result;
}

} // namespace HavenFSK
